### ISSUE TITLE
diagnosis of a specified area in simulation box

### ISSUE THREAD

**Issue by TrumeAAA** [Warwick-Plasma/epoch#515]
Hi,
   When we perform a 2D or even 3D simulation, the output file can be very large. Is it possible to only output E/B/ne...... of a specified area in the simulation box that we care about for example: a simution box of (0-100um)*(0-100um), and only output E/B/ne in the area of (50um-51um)*(50um-51um) ?
Thanks a lot.
All the best!

**Comment by Status-Mirror** [comment]
I believe EPOCH can do this using subsets. Check out the [Subsets of fields](https://epochpic.github.io/documentation/input_deck/input_deck_subset.html) section in the EPOCH documentation.

Hope this helps,
Stuart

**Comment by TrumeAAA** [comment]
Hi,
    Thank you very much for your reply. I tried to use Subset block to do it but failed. Details are as follows:
![control_block](https://github.com/Warwick-Plasma/epoch/assets/94948250/e2ac2cf7-d93a-44e1-ba4b-1066e8e0de90)
![subset_block](https://github.com/Warwick-Plasma/epoch/assets/94948250/1db78616-6fb9-4170-9a53-9a0516d5ac58)
![output_block](https://github.com/Warwick-Plasma/epoch/assets/94948250/9fa77681-0b40-487b-81a9-aa3f02b0b5fb)
    Here are my control block, subset blocks and output block. In this simulation, I tried to output electric field, current, number density and charge density of each species in a specified area defined in the subset blocks. However, the results in the SDF file looks weird. They are as follow ( Ey, Jy, ne, Qe for example ):
![Ey](https://github.com/Warwick-Plasma/epoch/assets/94948250/f2c13f62-df34-4e50-8e33-d2cc74e7a215)
![Jy](https://github.com/Warwick-Plasma/epoch/assets/94948250/e6938962-d688-47aa-92d5-e492dcf420f8)
![ne](https://github.com/Warwick-Plasma/epoch/assets/94948250/e28a13c3-5454-46da-9fa6-f40bd28cc862)
![Qe](https://github.com/Warwick-Plasma/epoch/assets/94948250/5c86b5fa-3705-4944-983a-bd8db23cc3b7)
    The value of each quantity goes to infinity. Maybe I made some mistakes while using this block and tried to solve for many times but failed.

**Comment by ktangtar** [comment]
I have seen this when doing subsets for fields as well, infinite values here and there.  The rest of the values should be correct though.  So I just delete those really large values.

**Comment by TrumeAAA** [comment]
> I have seen this when doing subsets for fields as well, infinite values here and there. The rest of the values should be correct though. So I just delete those really large values.

Sorry, I don't understand that. Suppose the results here is composed of correct values and large values. Since there is a distribution of the large values. So how did you do to make sure that you deleted all the large values while keeping the correct values?

**Comment by ktangtar** [comment]
> Sorry, I don't understand that. Suppose the results here is composed of correct values and large values. Since there is a distribution of the large values. So how did you do to make sure that you deleted all the large values while keeping the correct values?

iirc, the difference between the correct values and large values were about 100 orders of magnitude.  Based on your plots, it's probably true for you as well.  for example, if you say, all cells that are greater than 10^50 are set to nan, I think you probably will successfully delete all the large values and only keep the correct values.  At least, it was as easy as that for me.

**Comment by TrumeAAA** [comment]
> > Sorry, I don't understand that. Suppose the results here is composed of correct values and large values. Since there is a distribution of the large values. So how did you do to make sure that you deleted all the large values while keeping the correct values?
> 
> iirc, the difference between the correct values and large values were about 100 orders of magnitude. Based on your plots, it's probably true for you as well. for example, if you say, all cells that are greater than 10^50 are set to nan, I think you probably will successfully delete all the large values and only keep the correct values. At least, it was as easy as that for me.

Thanks. I tried to make the large values equal to 0 and the results looks normal. However, since real values at points with extremely large values are not 0, so the E, J, n and Q will looks discontinuous. Perhaps there are bugs to be fixed.

**Comment by Status-Mirror** [comment]
It sounds like there may be bugs to be fixed.

In EPOCH, you can have multiple output blocks for writing to multiple files. Can you verify that the fields in the subset cells do not match the fields in the equivalent cells of the full dump? If you can produce an input deck which demonstrates this behaviour (preferably for a small simulation), then I can mark this as a bug and get to debugging.

Cheers,
Stuart

**Comment by TrumeAAA** [comment]
Hi,
I performed a simple 2D simulation that is about the gamma ray generation in the interaction of an electron beam and a solid target(DOI: 10.1103/PhysRevLett.126.064801).
[input1.txt](https://github.com/Warwick-Plasma/epoch/files/11888572/input1.txt)
[input2.txt](https://github.com/Warwick-Plasma/epoch/files/11888573/input2.txt)
In the 1st input file, I tried to use 2 output blocks, one normal and another output only the subset area. However, it didn't work as I hope. The E, B field are only output normally without output of the subset area. For the n, Q, they are outputed without distinguishing their species. So I performed another simulation with output only in subset area. However the same thing happened for charge and number density. I dont know why.
Here is the comparison of the results between normal and subset output.
![normal_output](https://github.com/Warwick-Plasma/epoch/assets/94948250/b7f87f3b-0bbc-4390-9008-c345b481233c)
This is the result of the normal output.
![subset_output](https://github.com/Warwick-Plasma/epoch/assets/94948250/6daf5957-70b8-4246-aae4-c09888ea23e1)
This is the result when used subset block. There are some extremely large values that doesn't make sense. I simply set the values of the points with large values to be 0 in matlab:
**
ey(find(abs(ey)>1e40)) = 0;
bz(find(abs(bz)>1e40)) = 0;
jy(find(abs(jy)>1e40)) = 0;
n(find(abs(n)>1e40)) = 0;
**
And get
![subset_output](https://github.com/Warwick-Plasma/epoch/assets/94948250/f94dea7c-a457-47c1-ae43-81211ca73ebb)
It looks like the result of normal one. So the result is composed of correct values and large values when subset block is used.

**Comment by TrumeAAA** [comment]
Hi,
    I am not sure if this is enough to say that it is a bug. What is your opinion？
All the best!

**Comment by Status-Mirror** [comment]
This makes it seem like the problem may be machine-dependent. Just to make sure we are absolutely running the same code in the same way, could you perform these following steps for me:

Open a linux terminal, and run the commands:

```
mkdir debug_test
cd debug_test
git clone --recursive https://github.com/Warwick-Plasma/epoch.git
cd epoch/epoch2d
make -j4 COMPILER=gfortran DEFINE=-DPHOTONS
mkdir quick_test
cd quick_test
touch input.deck
```

Next, copy and paste the input deck I provided in my last message into your new input.deck file. Once this is done, you can run the simulation using:

```
dir=`pwd`
cd ..
mpirun -np 4 ./bin/epoch2d <<< $dir
```

This simulation took 7 seconds to run on my latptop. Finally, let us post-process the data to obtain heatmaps. I will provide instructions for plotting with MATLAB, as this is what I use for my work. Let me know if you use another post-processing package. It could be that the problem is in the Python post-processing scripts if this is what you're using?

For MATLAB, copy the relevant MATLAB processing scripts to the working directory

```
cp ../SDF/Matlab/* ./quick_test/
cd quick_test
matlab
```

Next, run the following commands in MATLAB to generate the plots

```
figure;
data = GetDataSDF('0015.sdf');
x_edges = data.Electric_Field.Ey.grid.x;
y_edges = data.Electric_Field.Ey.grid.y;
x_centres = 0.5*(x_edges(2:end) + x_edges(1:end-1));
y_centres = 0.5*(y_edges(2:end) + y_edges(1:end-1));
[x_plot, y_plot] = meshgrid(x_centres, y_centres);
surf(x_plot*1.0e6, y_plot*1.0e6, data.Electric_Field.Ey.data','EdgeColor','none');
view(2);
cbar = colorbar;

figure;
data = GetDataSDF('small0015.sdf');
x_edges = data.Electric_Field.Ey.Core_field_area.grid.x;
y_edges = data.Electric_Field.Ey.Core_field_area.grid.y;
x_centres = 0.5*(x_edges(2:end) + x_edges(1:end-1));
y_centres = 0.5*(y_edges(2:end) + y_edges(1:end-1));
[x_plot, y_plot] = meshgrid(x_centres, y_centres);
surf(x_plot*1.0e6, y_plot*1.0e6, data.Electric_Field.Ey.Core_field_area.data','EdgeColor','none');
view(2);
cbar = colorbar;
```

We now have two figures open in MATLAB, displaying the `Ey` output from the full and subset dumps in `0015.sdf` and `small0015.sdf` respectively. After some basic formatting (axis labels, text resizing), I have:

![both_results](https://github.com/Warwick-Plasma/epoch/assets/38282483/a71d485a-9d34-4740-a93a-d8c32ce4fb48)

As can be seen, there are no values exceeding |1e100| V/m in my test. Are you still finding issues with these plots, and could our differences be related to different plotting techniques?

Cheers,
Stuart

**Comment by TrumeAAA** [comment]
Hi @Status-Mirror,
    I tried it again ( smaller simulation case you provided ) using the latest version of EPOCH and get the correct results ! The version I used before is 4.17.16 and get large values here and there !
All the best!
H. Ai

**Comment by Status-Mirror** [comment]
Excellent news! I'm not sure which particular bug fix would have resolved your issue between 4.17.16 and 4.19.2, but I'm glad to hear the problem has been resolved.

I'll mark this issue as closed for now.

Cheers,
Stuart

**Comment by 2018-2025** [comment]
w1
